File: app/src/main/java/de/westnordost/streetcomplete/quests/postbox_collection_times/AddCollectionTimesForm.java
package de.westnordost.streetcomplete.quests.postbox_collection_times;

import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import java.util.ArrayList;

import javax.inject.Inject;

import de.westnordost.streetcomplete.Injector;
import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment;
import de.westnordost.streetcomplete.util.AdapterDataChangedWatcher;
import de.westnordost.streetcomplete.util.Serializer;


public class AddCollectionTimesForm extends AbstractQuestFormAnswerFragment
{
	public static final String TIMES = "times", NO_TIMES_SPECIFIED = "no_times_specified";

	private static final String TIMES_DATA = "times_data";

	private CollectionTimesAdapter collectionTimesAdapter;

	@Inject Serializer serializer;

	@Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
									   Bundle savedInstanceState)
	{
		View view = super.onCreateView(inflater, container, savedInstanceState);

		Injector.instance.getApplicationComponent().inject(this);

		View contentView = setContentView(R.layout.quest_collection_times);

		initCollectionTimesAdapter(contentView, savedInstanceState);

		Button addTimes = contentView.findViewById(R.id.btn_add);
		addTimes.setOnClickListener((v) -> collectionTimesAdapter.addNew());

		addOtherAnswer(R.string.quest_collectionTimes_answer_no_times_specified, () ->
		{
			new AlertDialog.Builder(getContext())
				.setTitle(R.string.quest_generic_confirmation_title)
				.setPositiveButton(R.string.quest_generic_confirmation_yes, (dialog, which) ->
				{
					Bundle answer = new Bundle();
					answer.putBoolean(NO_TIMES_SPECIFIED, true);
					applyAnswer(answer);
				})
				.setNegativeButton(R.string.quest_generic_confirmation_no, null)
				.show();
		});

		return view;
	}

	private void initCollectionTimesAdapter(View contentView, Bundle savedInstanceState)
	{
		ArrayList<WeekdaysTimes> data;
		if(savedInstanceState != null)
		{
			data = serializer.toObject(savedInstanceState.getByteArray(TIMES_DATA),ArrayList.class);
		}
		else
		{
			data = new ArrayList<>();
		}

		collectionTimesAdapter = new CollectionTimesAdapter(data, getContext(), getCountryInfo());
		collectionTimesAdapter.registerAdapterDataObserver(new AdapterDataChangedWatcher(this::checkIsFormComplete));
		RecyclerView collectionTimesList = contentView.findViewById(R.id.collection_times_list);
		collectionTimesList.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false));
		collectionTimesList.setAdapter(collectionTimesAdapter);
		collectionTimesList.setNestedScrollingEnabled(false);
		checkIsFormComplete();
	}

	@Override public void onSaveInstanceState(Bundle outState)
	{
		super.onSaveInstanceState(outState);
		outState.putByteArray(TIMES_DATA, serializer.toBytes(collectionTimesAdapter.getData()));
	}

	@Override protected void onClickOk()
	{
		Bundle answer = new Bundle();
		answer.putString(TIMES, collectionTimesAdapter.toString());
		applyAnswer(answer);
	}

	@Override public boolean isFormComplete() { return !collectionTimesAdapter.toString().isEmpty(); }
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/postbox_collection_times/AddCollectionTimesForm.kt
package de.westnordost.streetcomplete.quests.postbox_collection_times

import android.os.Bundle
import android.view.Menu
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.View
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.RecyclerView

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.databinding.QuestCollectionTimesBinding
import de.westnordost.streetcomplete.quests.AbstractQuestFormAnswerFragment
import de.westnordost.streetcomplete.quests.AnswerItem
import de.westnordost.streetcomplete.util.AdapterDataChangedWatcher
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json


class AddCollectionTimesForm : AbstractQuestFormAnswerFragment<CollectionTimesAnswer>() {

    override val contentLayoutResId = R.layout.quest_collection_times
    private val binding by contentViewBinding(QuestCollectionTimesBinding::bind)

    override val otherAnswers = listOf(
        AnswerItem(R.string.quest_collectionTimes_answer_no_times_specified) { confirmNoTimes() }
    )

    private lateinit var collectionTimesAdapter: CollectionTimesAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val viewData = loadCollectionTimesData(savedInstanceState)
        collectionTimesAdapter = CollectionTimesAdapter(viewData, requireContext(), countryInfo)
        collectionTimesAdapter.registerAdapterDataObserver( AdapterDataChangedWatcher { checkIsFormComplete() })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.collectionTimesList.layoutManager = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
        binding.collectionTimesList.adapter = collectionTimesAdapter
        binding.collectionTimesList.isNestedScrollingEnabled = false
        checkIsFormComplete()

        binding.addTimesButton.setOnClickListener { onClickAddButton(it) }
    }

    private fun onClickAddButton(v: View) {
        val rows = collectionTimesAdapter.collectionTimesRows

        val addTimeAvailable = rows.isNotEmpty()

        if (addTimeAvailable) {
            val popup = PopupMenu(requireContext(), v)
            if (addTimeAvailable) popup.menu.add(Menu.NONE, 0, Menu.NONE, R.string.quest_openingHours_add_hours)
            popup.menu.add(Menu.NONE, 1, Menu.NONE, R.string.quest_openingHours_add_weekdays)
            popup.setOnMenuItemClickListener { item ->
                when(item.itemId) {
                    0 -> collectionTimesAdapter.addNewHours()
                    1 -> collectionTimesAdapter.addNewWeekdays()
                }
                true
            }
            popup.show()
        } else {
            collectionTimesAdapter.addNewWeekdays()
        }
    }

    private fun loadCollectionTimesData(savedInstanceState: Bundle?): List<WeekdaysTimesRow> =
        savedInstanceState?.let { Json.decodeFromString(it.getString(TIMES_DATA)!!) } ?: listOf()

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(TIMES_DATA, Json.encodeToString(collectionTimesAdapter.collectionTimesRows))
    }

    override fun onClickOk() {
        applyAnswer(CollectionTimes(collectionTimesAdapter.createCollectionTimes()))
    }

    private fun confirmNoTimes() {
        AlertDialog.Builder(requireContext())
            .setTitle(R.string.quest_generic_confirmation_title)
            .setPositiveButton(R.string.quest_generic_confirmation_yes) { _, _ -> applyAnswer(NoCollectionTimesSign) }
            .setNegativeButton(R.string.quest_generic_confirmation_no, null)
            .show()
    }

    override fun isFormComplete() = collectionTimesAdapter.createCollectionTimes().isNotEmpty()

    companion object {
        private const val TIMES_DATA = "times_data"
    }
}
